### PR TITLE
feat: hot module replacement feature

### PR DESCRIPTION
**Pull request by elibolonur**
### 📖  Description

- Simple socket server implementation to enable HMR
- Webpack plugin is written for HMR support
- Add support for hot reloading and live reloading the content inside simulator both for Windows and iOS (Supports Vue based projects and Legacy projects)
- Vue watcher now uses @vue/cli-service API instead of executing command from bash
- Let user to choose a folder, if multiple folders found under Pitcher Folders which contains same `fileID`
- Automount Parallels VM as a  network drive (no need to manually mount it anymore, depends on static VM name for now)
- Search path is more precise on macOS
- Docs to be updated on js-platform repo

- [x] Tested on iOS
- [x] Tested on Windows
- [x] Tested with Vue projects 
- [x] Tested with Legacy projects

### CODE REVIEW

File: src/hmr/hmr-helper-legacy.js [hunk 1]
```diff
@@ -0,0 +1,193 @@
+/* eslint-disable prefer-template, prettier/prettier, no-var, consistent-return, object-shorthand, eqeqeq */
+
+/** Used for Legacy projects
+  * IMPORTANT NOTE: This file must to be ES5 compatible
```
Comment (kouts) [inline src/hmr/hmr-helper-legacy.js]: you can remove the `to` 😄

**Review by kouts — APPROVED** [review]
It's so good that it should be open-sourced.
Awesome work! 🥇

File: src/hmr/legacy-hmr.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  await injectHelperToHTML(destination)
+}
+
+// parses file name from full file path
+const parseFileName = (path) => path.substring(path.lastIndexOf('/') + 1)
+
+const emitChangesToClients = (wss, mode, changedFiles) => {
+  wss.clients.forEach((client) => {
+    if (mode === 'hot') {
+      const filesToEmit = Array.from(changedFiles).map((p) => parseFileName(p))
```
Comment (skix123) [inline src/hmr/legacy-hmr.js]: This should go before the loop, so it's done only once for each change

File: src/hmr/hmr-helper-legacy.js (hunk 1 above)
Comment (wojtiku) [reply src/hmr/hmr-helper-legacy.js]: To be, or not to be? 🤔

File: src/hmr/socket-server.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  const server = http.createServer(app)
+  const wss = new WebSocket.Server({ server })
+
+  // Example of wss listener
+  // wss.on('connection', (ws) => {
+  //   ws.on('message', (msg) => {
+  //     // received message
+  //     console.log('received: %s', msg)
+  //   })
+  // })
```
Comment (skix123) [inline src/hmr/socket-server.js]: Unnecessary code

File: src/utils/file-system.js [hunk 4]
```diff
@@ -55,15 +55,23 @@ const bashCopy = (sources, dest, ignored) => {
...
 
-      log('copied files', 'grey')
+      log('copied files', 'magenta')
       resolve()
     })
   })
 }
 
+// gets the folder name with parent folder name
+const getFolderNameWithparent = (path) => {
```
Comment (skix123) [inline src/utils/file-system.js]: `getFolderNameWithparent` -> `getFolderNameWithParent` (uppercase `Parent`)

File: src/hmr/socket-server.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+const startServer = (port) => {
+  const app = express()
+
+  app.get('/', (req, res) => {
+    res.set('Content-Type', 'text/html')
+    res.send(Buffer.from('<h2>Pitcher Watcher HMR</h2>'))
+  })
+
+  const server = http.createServer(app)
+  const wss = new WebSocket.Server({ server })
```
Comment (skix123) [inline src/hmr/socket-server.js]: `wss` means it's secure `ws` (like `http` vs `https`) - I think it's just `ws` here, right?

**Review by skix123 — COMMENTED** [review]
Great job! 👏

File: src/hmr/legacy-hmr.js (hunk 2 above)
Comment (elibolonur) [reply src/hmr/legacy-hmr.js]: Nice catch! Thanks

Comment (pegiadise) [reply src/hmr/legacy-hmr.js]: Then you also don't need the timeout? Or am I holding this wrong?

File: src/hmr/hmr-helper-legacy.js (hunk 1 above)
Comment (elibolonur) [reply src/hmr/hmr-helper-legacy.js]: `must be` 😄

**Review by pegiadise — APPROVED** [review]
Awesome. Gonna give it a spin! 🎉 🎊

File: src/hmr/legacy-hmr.js (hunk 2 above)
Comment (elibolonur) [reply src/hmr/legacy-hmr.js]: Timeout is to debounce the change event there actually. Lets assume you have changed 5 files and saved, chokidar is sending 5 events to the change hook which causes to execute post operations 5 times. Within 500ms delay, it is saving the changed file names and then finally executes the post operations.

File: src/hmr/socket-server.js (hunk 5 above)
Comment (elibolonur) [reply src/hmr/socket-server.js]: It was intended to be WebSocketServer, but ws is fine as well :) We could change it to `ws` tho if it seems unclear?

File: src/hmr/hmr-helper-legacy.js (hunk 1 above)
Comment (elibolonur) [reply src/hmr/hmr-helper-legacy.js]: fixed with 6221389

File: src/hmr/legacy-hmr.js (hunk 2 above)
Comment (elibolonur) [reply src/hmr/legacy-hmr.js]: fixed with e66007f

File: src/hmr/socket-server.js (hunk 3 above)
Comment (elibolonur) [reply src/hmr/socket-server.js]: fixed with 64e4bd5

File: src/utils/file-system.js (hunk 4 above)
Comment (elibolonur) [reply src/utils/file-system.js]: fixed with 3d92607

File: src/hmr/socket-server.js (hunk 5 above)
Comment (elibolonur) [reply src/hmr/socket-server.js]: fixed with 9cf53e6

**Review by skix123 — APPROVED** [review]

**Pull request merged**